sw/applications/spi_flash_write/spi_flash_write.c

```diff
@@ -0,0 +1,373 @@
+
+#ifndef USE_SPI_FLASH
+void handler_irq_fast_spi(void)
+{
+    // Disable SPI interrupts
+    spi_enable_evt_intr(&spi_host, false);
+    spi_enable_rxwm_intr(&spi_host, false);
+    // Clear fast interrupt
+    fast_intr_ctrl_t fast_intr_ctrl;
+    fast_intr_ctrl.base_addr = mmio_region_from_addr((uintptr_t)FAST_INTR_CTRL_START_ADDRESS);
+    clear_fast_interrupt(&fast_intr_ctrl, kSpi_e);
+    spi_intr_flag = 1;
+}
+#else
+void handler_irq_fast_spi_flash(void)
+{
+    // Disable SPI interrupts
+    spi_enable_evt_intr(&spi_host, false);
+    spi_enable_rxwm_intr(&spi_host, false);
+    // Clear fast interrupt
+    fast_intr_ctrl_t fast_intr_ctrl;
+    fast_intr_ctrl.base_addr = mmio_region_from_addr((uintptr_t)FAST_INTR_CTRL_START_ADDRESS);
+    clear_fast_interrupt(&fast_intr_ctrl, kSpiFlash);
+    spi_intr_flag = 1;
+}
+#endif
+
+void handler_irq_fast_dma(void)
+{
+    fast_intr_ctrl_t fast_intr_ctrl;
+    fast_intr_ctrl.base_addr = mmio_region_from_addr((uintptr_t)FAST_INTR_CTRL_START_ADDRESS);
+    clear_fast_interrupt(&fast_intr_ctrl, kDma_e);
+    dma_intr_flag = 1;
+}
+
+// Reserve memory array
+uint32_t flash_data[SPI_BYTES / 4] __attribute__ ((aligned (4))) = {0x76543210,0xfedcba98,0x579a6f90,0x657d5bee,0x758ee41f,0x01234567,0xfedbca98,0x89abcdef,0x679852fe,0xff8252bb,0x763b4521,0x6875adaa,0x09ac65bb,0x666ba334,0x44556677,0x0000ba98};
+uint32_t copy_data[SPI_BYTES / 4] __attribute__ ((aligned (4)))  = { 0 };
+
+int main(int argc, char *argv[])
+{
+    #ifndef USE_SPI_FLASH
+        spi_host.base_addr = mmio_region_from_addr((uintptr_t)SPI_START_ADDRESS);
+    #else
+        spi_host.base_addr = mmio_region_from_addr((uintptr_t)SPI_FLASH_START_ADDRESS);
+    #endif
+
+    // dma peripheral structure to access the registers
+    dma_t dma;
+    dma.base_addr = mmio_region_from_addr((uintptr_t)DMA_START_ADDRESS);
+
+    soc_ctrl_t soc_ctrl;
+    soc_ctrl.base_addr = mmio_region_from_addr((uintptr_t)SOC_CTRL_START_ADDRESS);
+
+    // Enable interrupt on processor side
+    // Enable global interrupt for machine-level interrupts
+    CSR_SET_BITS(CSR_REG_MSTATUS, 0x8);
+    // Set mie.MEIE bit to one to enable machine-level fast spi interrupt
+    #ifndef USE_SPI_FLASH
+        const uint32_t mask = 1 << 20;
+    #else
+        const uint32_t mask = 1 << 21;
+    #endif
+    CSR_SET_BITS(CSR_REG_MIE, mask);
+    spi_intr_flag = 0;
+
+    // Select SPI host as SPI output
+    soc_ctrl_select_spi_host(&soc_ctrl);
+    // Enable SPI host device
+    spi_set_enable(&spi_host, true);
+
+    // SPI and SPI_FLASH are the same IP so same register map
+    uint32_t *fifo_ptr = spi_host.base_addr.base + SPI_HOST_DATA_REG_OFFSET;
+
+    // -- SPI CONFIGURATION -- 
+    // Configure SPI clock
+    // SPI clk toggles at half of the freq as the core clk when clk_div = 1
+    // SPI_CLK = CORE_CLK/(2 + 2 * CLK_DIV) < CLK_MAX => CLK_DIV > (CORE_CLK/CLK_MAX - 2)/2
+    const uint16_t clk_div = (REFERENCE_CLOCK_Hz/(1000*1000*FLASH_CLK_MAX) - 2)/2 + 1;
```

Comment:
Nice change. Just take the clock from the soc_ctrl peripheral as the REFERENCE_CLOCK_Hz might not be the current frequency (e.g., in an actual chip with an FLL, REF_CLOCK is usually 32768Hz and use to create a faster clock)